### ISSUE TITLE
Missing features in CharmmPsfFile.createSystem()

### ISSUE THREAD

**Issue by Olllom** [openmm/openmm#2353]
While working with CHARMM input files in OpenMM, I have come across several shortcomings of the PSF parser `CharmmPsfFile.createSystem()`. I am willing to fix this soon, but before doing so, I wanted to check if anybody else has suggestions.

Here is what I found annoying:
- Although most of the CHARMM force field is parameterized without analytic dispersion correction, there is currently no option in `createSystem()` to manually disable the correction. Instead, the user has to disable it manually by walking through all `NonbondedForce` and `CustomNonbondedForce` objects of the system.
- `createSystem()` sometimes merges residues that have the same RESID from different segments. 
- There is an annoying warning "Detected PSF molecule section that is WRONG. Resetting molecularity," that gets thrown even for some simple PSF files that CHARMM does not have any trouble with. 

CharmmCrdFile:
- It might be useful to (maybe by default??) shift all coordinates by half a box length. Most CHARMM systems are build around the origin and we usually have to rewrap coordinates by hand. Not such a big deal, but anyway.

@jing-huang 
@andysim Reminder to modularize `createSystem` for your plugin.

**Comment by peastman** [comment]
Fixes for these would be great!

Does CHARMM have the ability to apply a dispersion correction?  If so, how (and where) do you specify whether to use it?  Anyway, the default behavior of CharmmPsfFile should definitely be to match what CHARMM would do when processing the same file.

**Comment by Olllom** [comment]
CHARMM has a dispersion correction that gets activated by the `LRC` or `LRC_MS` keywords in NBONDS: https://www.charmm.org/charmm/documentation/by-version/c42b1/params/doc/nbonds/ 
When using the CHARMM force fields, especially for lipid bilayers, the LRC should not be turned on.

One concern regarding backwards compatibility is CHARMM-GUI, which uses the parsers in its OpenMM input files.

**Comment by peastman** [comment]
So `createSystem()` should be able to tell whether the correction is supposed to be enabled or not?  Then it can just automatically do the right thing.

> One concern regarding backwards compatibility is CHARMM-GUI, which uses the parsers in its OpenMM input files.

Do the scripts it generates disable the correction?  If so, this won't affect them.  If not, this will fix a bug.

**Comment by Olllom** [comment]
> So createSystem() should be able to tell whether the correction is supposed to be enabled or not? Then it can just automatically do the right thing.

In my opinion, LRC should be disabled by default and there should be a keyword argument to enable it but @jing-huang may think differently.

> Do the scripts it generates disable the correction? If so, this won't affect them. If not, this will fix a bug.

Excellent point. Actually, I do not see it being disabled in some older CHARMM-GUI files that I had lying around, which is surprising. I will regenerate them later and see if something has changed recently. If not we can check with them if this is a feature or a bug.

It is for sure though that they manually rewrap the coordinates coming from the CRD file.

**Comment by ajsilveira** [comment]
Hi all. I have a question related to this issue, since I am trying to simulate a system obtained by using CHARMM-GUI: Why is the coordinate shifting necessary? As the system is subject to periodic boundary conditions, I thought that it didn't matter.

**Comment by Olllom** [comment]
> As the system is subject to periodic boundary conditions, I thought that it didn't matter.

In principle, it does not matter, but
- You usually do not want to break molecules across periodic boundary conditions. For codes that use domain decomposition, this may result in a performance penalty.
- Postprocessing trajectories that have the water phase in the center and two leaflets of the membrane on the outside can be a little awkward. It is unusual, at the very least.

**Comment by peastman** [comment]
> You usually do not want to break molecules across periodic boundary conditions. For codes that use domain decomposition, this may result in a performance penalty.

For OpenMM it makes no difference at all.

> Postprocessing trajectories that have the water phase in the center and two leaflets of the membrane on the outside can be a little awkward. It is unusual, at the very least.

This doesn't really solve the problem.  You can arrange that your solute starts in the middle of the box, but after 100 ns who knows where it will have diffused to?  That's why analysis programs like MDTraj and CPPTraj have functions to translate them back again.  That's often one of the first steps in analysis pipelines.

**Comment by jing-huang** [comment]
Yes having an option in createSystem() to turn on and off analytic LRC would be really important, and I actually agree that the default option should be "off" in order to be compatible with CHARMM force field, in particular lipid force field. 

I don't see any necessity to modify CharmmCrdFile. Shifting the coordinates at the beginning of the simulation won't make analysis easier at all. 

"Detected PSF molecule section that is WRONG. Resetting molecularity," should be disabled. If I remembered correctly, this section in charmm psf is used by the CHARMM fluctuating charge force field, which OpenMM doesn't support anyway.

**Comment by peastman** [comment]
We've also been discussing this offline over email.  I think we've got a plan for what to do:

1. The dispersion correction will be either enabled or disabled based on the LRC or LRC_MS keyword in the parameter file.  (This is how CHARMM decides whether to include it.)

2. We'll leave the crd reader as is (not shift the coordinates).

3. We'll get rid of the unnecessary warning.

4. We'll fix the problem with it incorrectly merging residues.

@Olllom if you want to go ahead and implement this, that would be awesome.

**Comment by Olllom** [comment]
@peastman I will work on it

> 1. The dispersion correction will be either enabled or disabled based on the LRC or LRC_MS keyword in the parameter file.  (This is how CHARMM decides whether to include it.)

Sorry for being annoying but one final remark here.
CHARMM has several ways of specifiying the settings for nonbonded interactions, like cutoffs, switch distances, and also long-range corrections. The parameter files contain statements that specify the default settings
```
NONBONDED nbxmod  5 atom cdiel fshift vatom vdistance vfswitch -
cutnb 14.0 ctofnb 12.0 ctonnb 10.0 eps 1.0 e14fac 1.0 wmin 1.5
```
In practice, when running a CHARMM simulation, these settings are often updated by calling `ENERGY` (or `UPDATE`) in the run script

```
UPDATE cutim 14. imgfrq -1 cutnb 14. inbfrq -1 -
  bycb atom vatom vfswitch ctofnb 12. ctonnb 8. -
  ewald pme spline order 6 fftx @FX ffty @FX fftz @FZ kappa 0.32
```
(This, for example, will set the switch distance to 8 Angstrom, which is how much of the lipid force field was parameterized.)

A setting like `LRC` would usually not be specified in the parameter file but in this `UPDATE` command (or, equivalently, the `ENERGY` or `DYNA` command).

As far as I can see, the `openmm.app.CharmmParameterSet` does not even bother parsing the arguments of the `NONBONDED` command from the parameter files. Instead, cutoff and switching distance, PME settings, etc. are specified as arguments of `createSystem`. This made me think, that the long-range correction would belong there, too.

**Comment by peastman** [comment]
This is a philosophical question.  Do you consider the dispersion correction to be part of the force field definition, or is it a user option where people might reasonably want different choices in different situations?  What I've been hearing from the CHARMM developers is that they view it as part of the force field definition.  The force field was parametrized to give correct densities without the correction, and they would consider it an error to ever use the force field with a correction.

Note a difference from the cutoff distance and similar settings: this doesn't involve any tradeoff between speed and accuracy.  The force field was parametrized with a cutoff of 12A, but someone might reasonably choose to use a smaller cutoff to increase performance.  By doing that they would make the results less accurate, but in some situations that's a fair trade to get more sampling.  But the dispersion correction adds negligible overhead.  There's no reason ever to turn it off unless you think it makes your results less accurate; and if it makes the results less accurate, there's no reason ever to turn it on.

I assume future versions of the CHARMM force fields will be parametrized with LJPME, and that will make the whole question moot in the same way PME already has for Coulomb interactions.  With PME, there's no need for corrections and no "right" cutoff to use.  The cutoff just shifts work between real space and reciprocal space.  You can get effectively identical results with any cutoff.

**Comment by peastman** [comment]
This was fixed by #2359.
